[ALPHA] V2 Alpha Blog Post (#5343)
pydantic/pydantic: docs/blog/pydantic-v2-alpha.md

# Pydantic V2 Pre Release

<aside class="blog" markdown>
![Terrence Dorsey](/img/terrencedorsey.jpg)
<div markdown>
  **Terrence Dorsey & Samuel Colvin** &bull;&nbsp;
  [:material-github:](https://github.com/pydantic) &bull;&nbsp;
  [:material-twitter:](https://twitter.com/pydantic) &bull;&nbsp;
  :octicons-calendar-24: April 3, 2023 &bull;&nbsp;
  :octicons-clock-24: 8 min read
</div>
</aside>

---

We're excited to announce the first alpha release of Pydantic V2!

> This first Pydantic V2 alpha is no April Fool's joke, for a start we missed our April 1st target date :cry:.

Reviewer: ```suggestion
This first Pydantic V2 alpha is no April Fool's joke &mdash; for a start we missed our April 1st target date :cry:.
```

See [the docs](/) for examples of Pydantic at work.

## Getting started with the Pydantic V2 alpha

Your feedback will be a critical part of ensuring that we have made the right tradeoffs with the API changes in V2.

To get started with the Pydantic V2 alpha, install it from PyPI. 
We recommend using a virtual environment to isolate your testing environment:

```bash
pip install --pre -U "pydantic>=2.0"
```

Note that there are still some rough edges and incomplete features, and while trying out the Pydantic V2 alpha releases you may experience errors. 
We encourage you to try out the alpha releases in a test environment and not in production. 
Some features are still in development, and we will continue to make changes to the API.

If you do encounter any issues, please [create an issue in GitHub](https://github.com/pydantic/pydantic/issues) using the `bug V2` label. 
This will help us to actively monitor and track errors, and to continue to improve the library’s performance.

This will be the first of several upcoming alpha releases. As you evaluate our changes and enhancements, 
we encourage you to share your feedback with us.

Please let us know:

> * If you don't like the changes, so we can make sure pydantic remains a library you enjoy using.

Reviewer: ```suggestion
* If you don't like the changes, so we can make sure Pydantic remains a library you enjoy using.
```

> * If this breaks your usage of pydantic so we can fix it, or at least describe a migration path.

Reviewer: ```suggestion
* If this breaks your usage of Pydantic so we can fix it, or at least describe a migration path.
```

Thank you for your support, and we look forward to your feedback.

---

## Headlines

Here are some of the most interesting new features in the current Pydantic V2 alpha release. 
For background on plans behind these features, see the earlier [Pydantic V2 Plan](/blog/pydantic-V2/) blog post.

The biggest change to Pydantic V2 is [`pydantic-core`](https://github.com/pydantic/pydantic-core) &mdash;
all validation logic has been rewritten in Rust and moved to a separate package, `pydantic-core`.
This has a number of big advantages:

* **Performance** - Pydantic V2 is 5-50x faster than Pydantic V1.

> * **Safety** - Pydantic V2 should have far fewer bugs long term.

Reviewer 2: Copy edit suggestion (feel free to ignore if you don't like it): 

> We've made changes to the architecture that we think will help us maintain Pydantic V2 with far fewer bugs in the long term.

Or similar?

Reviewer 2: And I just realized this bullet is basically the same as the next one. Feels like it could use a bit of a rework perhaps?

> * **Safety** - Pydantic V2 should have far fewer bugs long term.

Reviewer: ```suggestion
* **Safety & maintainability** - We've made changes to the architecture that we think will help us maintain Pydantic V2 with far fewer bugs in the long term.
```

> * **Maintainability** - we're hoping that, long term, V2 is easier to extend without the risk of introducing bugs or breaking changes.

Reviewer: ```suggestion
```

With the use of `pydantic-core`, the majority of the logic in the Pydantic library is dedicated to generating
"pydantic core schema" &mdash; the schema used define the behaviour of `pydantic-core` validators and serializers.  

### Ready for experimentation

* **BaseModel** - the core of validation in Pydantic V1 remains, albeit with new method names.

> * **dataclasses** - Pydantic dataclasses are improved and ready to test.

Reviewer 2: ```suggestion
* **Dataclasses** - Pydantic dataclasses are improved and ready to test.
```

The capitalization feels a bit inconsistent here. I recognize the module is called `dataclasses` but I think capitalizing is better here.

Same for "**strict mode**" below — I'd personally use "**Strict mode**"

* **Serialization** - dumping/serialization/marshalling is significantly more flexible, and ready to test.

> * **strict mode** - one of the biggest additions in Pydantic V2 is strict mode, which is ready to test.

Reviewer: ```suggestion
* **Strict mode** - one of the biggest additions in Pydantic V2 is strict mode, which is ready to test.
```

* **JSON Schema** - generation of JSON Schema is much improved and ready to test.
* **Generic Models** - are much improved and ready to test.
* **Recursive Models** - and validation of recursive data structures is much improved and ready to test.
* **Custom Types** - custom types have a new interface and are ready to test.
* **Custom Field Modifiers** - used via `Annotated[]` are working and in use in Pydantic itself. 
* **Validation without a BaseModel** - the new `AnalyzedType` class allows validation without the need for a `BaseModel` class, and it's ready to test.
* **TypedDict** - we now have full support for `TypedDict` via `AnalyzedType`, it's ready to test.

### Still under construction

* **Documentation** - we're working hard on full documentation for V2, but it's not ready yet.
* **Conversion Table** - a big addition to the documentation will be a conversion table showing how types are coerced, this is a WIP.

> * **BaseSettings** - `BaseSettings` will move to a separate `pydantic-settings` package, it's not yet ready to test.

Reviewer 2: @samuelcolvin we probably need to give people a suggestion of what to do, since `BaseSettings` isn't in pydantic v2. If they are using it in their codebase, I think they won't be able to test out the alpha.

* **validate_arguments** - the `validate_arguments` decorator remains and is working, but hasn't been updated yet.
* **Hypothesis Plugin** - the Hypothesis plugin is yet to be updated.
* **computed fields** - we know a lot of people are waiting for this, we will include it in Pydantic V2.
* **Error messages** - could use some love, and links to docs in error messages are still to be added.
* **Migration Guide** - we have some pointers below, but this needs completing.

## Migration Guide

**Please note:** this is just the beginning of a migration guide. We'll work hard up to the final release to prepare
a full migration guide, but for now the following pointers should be some help while experimenting with V2.

### Changes to BaseModel

* Various method names have been changed; `BaseModel` methods all start with `model_` now. 
  Where possible, we have retained the old method names to help ease migration, but calling them will result in `DeprecationWarning`s.
  * Some of the built-in data loading functionality has been slated for removal. 
    In particular, `parse_raw` and `parse_file` are now deprecated. You should load the data and then pass it to `model_validate`.

> * The `from_orm` method has been removed; you can now just use `model_validate` (equivalent to `parse_obj` from v1) to achieve something similar, 

Reviewer: ```suggestion
* The `from_orm` method has been removed; you can now just use `model_validate` (equivalent to `parse_obj` from Pydantic V1) to achieve something similar, 
```

  as long as you've set `from_attributes=True` in the model config.
* The `__eq__` method has changed for models; models are no longer considered equal to the dicts.
* Custom `__init__` overrides won't be called. This should be replaced with a `@root_validator`.
* Support for `__root__` models has been removed. However, you can effectively achieve something equivalent through the use of 

>   `@root_validator`, `@model_serializer`, and `__pydantic_modify_json_schema__`. More details will be available soon.

Reviewer 2: @samuelcolvin any thoughts on how to provide these extra details? (This is basically ready now, but not quite sure how to explain; doesn't seem appropriate to include the full set of functions to define in-line here..)

Reviewer 2: I'm tentatively planning to add a permalink to the implementation in one of the tests that uses a root model (after #5340 is merged), rather than an issue/discussion, and wait for someone to ask if they have questions/issues. But I could do that instead.

Reviewer 2: See https://github.com/pydantic/pydantic/pull/5357 for an update to this

Author: #5357 lgtm.

### Changes to Pydantic Dataclasses

* The `__post_init__` in Pydantic dataclasses will now be called after validation, rather than before.
* We no longer support `extra='allow'` for Pydantic dataclasses, where extra attributes passed to the initializer would be 
  stored as extra fields on the dataclass. `extra='ignore'` is still supported for the purposes of allowing extra fields while parsing data; they just aren't stored.
* `__post_init_post_parse__` has been removed.
* Nested dataclasses no longer accept tuples as input, only dict.

### Changes to Config

* To specify config on a model, it is now deprecated to create a class called `Config` in the namespace of the parent `BaseModel` subclass. 
  Instead, you just need to set a class attribute called `model_config` to be a dict with the key/value pairs you want to be used as the config.
  
The following config settings have been removed:

> * `allow_mutation`

Reviewer: ```suggestion
* `allow_mutation`.
```

> * `error_msg_templates`

Reviewer: ```suggestion
* `error_msg_templates`.
```

* `fields` — this was the source of various bugs, so has been removed. You should be able to use `Annotated` on fields to modify them as desired.
* `getter_dict` — `orm_mode` has been removed, and this implementation detail is no longer necessary.
* `schema_extra` — you should now use the `json_schema_extra` keyword argument to `pydantic.Field`.
* `smart_union`.

> * `underscore_attrs_are_private` — the Pydantic V2 behavior is now the same as if this was always set to True in Pydantic V1.

Reviewer: ```suggestion
* `underscore_attrs_are_private` — the Pydantic V2 behavior is now the same as if this was always set to `True` in Pydantic V1.
```

The following config settings have been renamed:

* `allow_population_by_field_name` → `populate_by_name`
* `anystr_lower` → `str_to_lower`
* `anystr_strip_whitespace` → `str_strip_whitespace`
* `anystr_upper` → `str_to_upper`
* `keep_untouched` → `ignored_types`
* `max_anystr_length` → `str_max_length`
* `min_anystr_length` → `str_min_length`
* `orm_mode` → `from_attributes`
* `validate_all` → `validate_default`

### Changes to Validators

* Raising a `TypeError` inside a validator no longer produces a `ValidationError`, but just raises the `TypeError` directly. 
  This was necessary to prevent certain common bugs (such as calling functions with invalid signatures) from 
  being unintentionally converted into `ValidationError` and displayed to users. 
  If you really want `TypeError` to be converted to a `ValidationError` you should use a `try: except:` block that will catch it and do the conversion.

> * `each_item` validators are deprecated and should be replaced with a type annotation using Annotated to apply a validator 

Reviewer: ```suggestion
* `each_item` validators are deprecated and should be replaced with a type annotation using `Annotated` to apply a validator 
```

* `each_item` validators are deprecated and should be replaced with a type annotation using `Annotated` to apply a validator 
  or with a validator that operates on all items at the top level.
* Changes to `@validator`-decorated function signatures.
* The `stricturl` type has been removed.
* Root validators can no longer be run with `skip_on_failure=False`.

### Changes to Validation of specific types

* Integers outside the valid range of 64 bit integers will cause `ValidationError`s during parsing. 
  To work around this, use an `IsInstance` validator (more details to come).
* Subclasses of built-ins won't validate into their subclass types; you'll need to use an `IsInstance` validator to validate these types.

### Changes to Generic models

* While it does not raise an error at runtime yet, subclass checks for parametrized generics should no longer be used. 
  These will result in `TypeError`s and we can't promise they will work forever. However, it will be okay to do subclass checks against _non-parametrized_ generic models

### Other changes

* `GetterDict` has been removed, as it was just an implementation detail for `orm_mode`, which has been removed.

### AnalyzedType

> PydanticV1 treated non-`BaseModel` types as second class citizens.

Author: ```suggestion
Pydantic V1 didn't have good support for validation or serializing non-`BaseModel`.
```

To work with them you had to create a "root" model or use the utility functions in `pydantic.tools` (`parse_obj_as` and `schema_of`).

> In Pydantic V2 this is _a lot_ easier: the `AnalyzedType` class lets you build an object that behaves almost like a `BaseModel`class which you can use for a lot of the use cases of root models and as a complete replacement for `parse_obj_as` and `schema_of`.

Author: ```suggestion
In Pydantic V2 this is _a lot_ easier: the `AnalyzedType` class lets you build an object that behaves almost like a `BaseModel` class which you can use for a lot of the use cases of root models and as a complete replacement for `parse_obj_as` and `schema_of`.
```

### AnalyzedType

> In Pydantic V1 if you wanted to validate a non-`BaseModel` type (like `Dict[str, int]` or `List[SomeModel]`) you had to use a "root" model.

Reviewer 2: There was also `tools.parse_obj_as`, which was added specifically for this purpose, and had a pretty clean API (imo). The important reasons to drop it were:
* It was inefficient — building the parsing model as part of the call (though admittedly it had an lru_cache with cache size 2048, so maybe not super likely to hit that).
* It wasn't super compatible with our new approach to serialization.

I guess you could argue that in some sense, the `pydantic.tools._get_parsing_type` has been replaced by `AnalyzedType` which does all of that and more. But considering that's more of an implementation detail I don't think that needs to be called out to users.

I _do_ think we should mention `pydantic.tools.parse_obj_as` here though.

Reviewer 2: (I'll also note that `pydantic.tools` had a function for getting the json schema of arbitrary types as well, so it really seems to me that the main benefit over the old approach is the serialization stuff, and any possible lru_cache-induced issues)

Reviewer 2: To that end, might make sense to say something about how we had utility functions for implementing this workflow for `__root__` models before, and because it is now simplified we have deprecated those functions. And you can replace `parse_obj_as(Dict[int, List[str], data)` with `AnalyzedType(Dict[int, List[str]]).validate_python(data)`, etc.

Reviewer 3: Do we need to explain why we're replacing `parse_obj_as` with this? Or can I just say "this replaces `pydantic.tools.parse_obj_as` and `pydantic.tools.json_schema` (or whatever the latter was called)

Author: I think we're keeping those types, just deprecated, and offering a 1 to 1 replacement, so that's fine.

The other big changes is that there's no `BaseModel` in the background, so performance should increased markedly.
